sharedconf/extension.py

```diff
@@ -0,0 +1,85 @@
+from pathlib import Path
+
+from sphinx import version_info
+from sphinx.util import logging
+
+from . import __version__, utils
+
+log = logging.getLogger(__name__)
+
+
+def _get_docset_from_config(config):
+    docset = utils.get_docset(config.sharedconf_docsets, config.sharedconf_env_var)
+    return docset, config.sharedconf_docsets[docset]
+
+
+def _override_srdir(app, config):
+    """
+    Override app.srcdir with the value from the current docset.
+
+    If the docset doesn't have a path,
+    the name of the docset is used as the path.
+
+    Values can be absolute or relative to the original source dir.
+    """
+    docset, options = _get_docset_from_config(config)
+    log.info("Using docset: %s", docset)
+
+    original_srcdir = Path(app.srcdir)
+    new_srcdir = Path(options.get("path", docset))
+    if not new_srcdir.is_absolute():
+        new_srcdir = original_srcdir / new_srcdir
+    app.srcdir = str(new_srcdir.resolve())
+
+
+def _override_config(config):
+    """Override the config with specific values from the current docset."""
+    docset, options = _get_docset_from_config(config)
+    docset_config = options.get("config", {})
+    # Taken from
+    # https://github.com/sphinx-doc/sphinx/blob/586de7300a5269de93b4d813408ffdbf10ce2409/sphinx/config.py#L222-L222
+    pre_init_options = {"needs_sphinx", "suppress_warnings", "language", "locale_dirs"}
+    for option, value in docset_config.items():
```

Comment:
This method of overriding the configuration file seems redundant -- the user would already have independent values in `docs/project1/conf.py` and `docs/project2/conf.py`. Could we get away with doing an import of most of these values, instead of duplicating this data in a common config file?

I'm sure users will just want to do a:

```python

import projecta.conf as projecta_conf
import projectb.conf as projectb_conf

sharedconf_docsets = {
    "projecta": {"config": {
        "project": proejcta.project,
        "version": projecta.version,
        ...
    }
}

Reply:
I like the idea of having the settings in different files and also have the possibility of using just one file (if the user only needs to override a couple of values on each project).

I see this being used in different ways:

```python
from multiproject.utils import get_project

extensions = [
    "sharedconf.extension",
    "sphinx.ext.intersphinx",
    "sphinx.ext.autodoc",
]

master_doc = "index"
language = "en"

multiproject_projects = {
    "api": {},
    "dev": {
        "config": {
            # Is this way of setting the options still useful?
            # Could be for simple projects,
            # so users don't need to import stuff.
            "project": "Dev project",
        },
    },
    "user": {},
    'auto': {
        # This is almost the same as using `config`,
        # but we set the values automatically from the `auto/conf.py` file?.
        # This has the same restrictions as using `config`.
        'config_file': 'auto/conf.py',
    },
}

current_project  = get_project(multiproject_projects)

# Set all values directly
# -----------------------

if current_project == 'api':
    from api.conf import *
elif current_project == 'user':
    from user.conf import *

# Set value by value
# ------------------

import api.conf as api_conf
import user.conf as user_conf
if current_project == 'api':
    config = api_conf
elif current_project == 'user':
    config =  user_conf

# Replace the original values
project = config.project
version = config.version
# Extending the original value
extensions += config.extensions
```

Reply:
> and also have the possibility of using just one file

Ahh I see where you're coming from. We could certainly support both. I do think the auto import method would be used the most, so I suppose I'm just advocating for adding that now/soon.

My guess would be that users would still maintain 2 full sphinx subprojects for more standard local development, so could see the `multiproject_projects['dev']` being an advanced use case.

I did like how your original example had all the configuration that a project would need in one place, instead of requiring manual `import` from the user. Manual set up could be an advanced usage, but we would recommend something like your `multiproject_projects['auto']` for almost all use cases.